File: Stack/Problem150.cpp
/**
 * https://leetcode.com/problems/evaluate-reverse-polish-notation/
 */

#include <iostream>
#include <vector>
#include <stack>
#include <bits/stdc++.h>
#include <string>

using namespace std;

class Solution {
public:
    bool check_int(string s) {
        for(int  i = 0; i < s.length(); i++) {
            if(s[0] == '-' && s.length() > 1) {
                continue;
            }

            if(isdigit(s[i])) {
                continue;
            } else {
                return false;
            }
        }
        return true;
    }

    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        for(auto& s : tokens) {
            if(check_int(s)) {
                int num = stoi(s);
                st.push(num);
            } else if(s == "+" && st.size() > 1) {
                int first = st.top();
                st.pop();
                int second = st.top();
                st.pop();
                st.push(first + second);
            } else if(s == "-" && st.size() > 1) {
                int first = st.top();
                st.pop();
                int second = st.top();
                st.pop();
                st.push(second - first);
            } else if(s == "*" && st.size() > 1) {
                int first = st.top();
                st.pop();
                int second = st.top();
                st.pop();
                st.push(first * second);
            } else if(s == "/" && st.size() > 1) {
                int first = st.top();
                st.pop();
                int second = st.top();
                st.pop();
                if (first == 0) {
                    throw runtime_error("Division by zero error");
                }
                st.push(second / first);
            }
            // cout << st.top() << endl;
        }
        return st.top();
    }
};

int main(void) {
    Solution sol;
    vector<string> str = {"4","3","-"};
    sol.evalRPN(str);
}

File: Stack/Python/problem150.py
class Solution:
    def calculate(self, n1, n2, op):
        if op == "+":
            return n1 + n2
        if op == "-":
            return n1 - n2
        if op == "*":
            # print(n1)
            return n1 * n2
        if op == "/":
            # print(n1 // n2)
            res = n1 / n2
            if res < 0:
                return math.ceil(res)
            return n1 // n2
    def evalRPN(self, tokens: List[str]) -> int:
        st = []
        for c in tokens:
            if c.lstrip("-").isdigit():
                st.append(int(c))
            else:
                if len(st) < 2:
                    return
                n2 = st.pop()
                n1= st.pop()
                # print("%d %d %d" %(n1, n2, self.calculate(n1, n2, c)))
                st.append(self.calculate(n1, n2, c))
        if len(st) != 1:
            return
        return st[-1]
        